compiler/rustc_hir_analysis/src/astconv/mod.rs

```diff
@@ -2332,13 +2331,115 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         self.ast_ty_to_ty_inner(ast_ty, false, true)
     }
 
+    fn check_delegation_constraints(&self, sig_id: DefId, span: Span, emit: bool) -> bool {
+        #[inline]
+        fn generics_param_count(param_count: ty::GenericParamCount) -> usize {
+            param_count.lifetimes + param_count.types + param_count.consts
+        }
+
+        let mut error_occured = false;
+        let sig_span = self.tcx().def_span(sig_id);
+        let mut try_emit = |descr| {
+            if emit {
+                self.tcx().dcx().emit_err(crate::errors::NotSupportedDelegation {
+                    span,
+                    descr,
+                    callee_span: sig_span,
+                });
+            }
+            error_occured = true;
+        };
+
+        if let Some(node) = self.tcx().hir().get_if_local(sig_id)
+            && let Some(decl) = node.fn_decl()
+            && let hir::FnRetTy::Return(ty) = decl.output
+            && let hir::TyKind::InferDelegation(_, _) = ty.kind
+        {
+            try_emit("recursive delegation");
+        }
+
+        let sig = self.tcx().fn_sig(sig_id).instantiate_identity();
+        if sig.output().has_opaque_types() {
+            try_emit("delegation to a function with opaque type");
+        }
+
+        let res_generics = self.tcx().generics_of(sig_id);
+        let parent = self.tcx().parent(self.item_def_id());
+        let parent_generics = self.tcx().generics_of(parent);
+        let self_ty_bias = (self.tcx().def_kind(parent) == DefKind::Trait) as usize;
+        if parent_generics.count() + generics_param_count(res_generics.own_counts()) > self_ty_bias
```

Comment:
this is simply `res_generics.count()`, is it not?

Reply:
Consider the example
```
trait Trait<T0> {
    fn foo<T1>() {}
}

struct Struct<T2> { /* ... */ }

impl<T0, T2> Trait<T0> for Struct<T2>  {
    /* reuse Trait::foo */
}
```
For a delegation item in a trait impl `sig_id` is a trait method resolution. So, `res_generics = [Self, T1]`, `parent_generics = [T0, T2]`. Expected generics for the `foo`: `[T0, T1, T2]`. That's where the addition comes from.

Reply:
My mistake. Conditions have been replaced.